### PR TITLE
Get quickplay playtime from Overwatch profile

### LINKED ISSUE #6: Feature request: Playtime for Overwatch
I'd like to see my Overwatch playtime in the Galaxy Launcher.

The playtime can be grabbed from the official Overwatch player profiles or a fanmade API.
Playtime is divided into quick play and ranked play - and have to be added to get a total play time.

## Official player profiles

URL: https://playoverwatch.com/en-us/career/pc/[username]-[number]

_Example: https://playoverwatch.com/en-us/career/pc/saitho-2703_

Total playtime for Quick play is located in the category "Game" as "Time played" (HH:MM:SS).
Playtime for competitive play is located at the same place, but one has to switch to competitive stats "tab" above the statistics.

## Fanmade API

URL: https://ow-api.com/v1/stats/pc/eu/[username]%23[number]/complete

_Example: https://ow-api.com/v1/stats/pc/eu/saitho%232703/complete_

Total playtime is located in `quickPlayStats.careerStats.allHeroes.game.timePlayed` and `competitiveStats.careerStats.allHeroes.game.timePlayed`.

### PR DESCRIPTION
**Pull request by saitho**
As title. Resolves #6.

However I wasn't able to run the tests via _inv_ as my system can't find the package even though I installed it. :/

I will also try to fetch Overwatch achievements in a separate PR.

### CODE REVIEW

File: src/plugin.py [hunk 1]
```diff
@@ -30,6 +30,16 @@
...
 class BNetPlugin(Plugin):
+    async def get_game_time(self, game_id, context):
+        qpTimeMinutes = 0
+        log.info('game_id: ' + game_id)
+        if game_id == "5272175": # Overwatch
+            player_data = await self.backend_client.get_ow_player_data()
+            qpTime = re.search('0x0860000000000026.+?(?:(?P<h>\d+):)?(?P<m>\d+)', player_data.text)
+            qpTimeMinutes = int(qpTime.group('h')) * 60 + int(qpTime.group('m'))
+            log.info('Overwatch quickplay playtime [minutes]: ' + str(qpTimeMinutes))
+        return GameTime(game_id, qpTimeMinutes, None)
```
Comment (elkangaroo) [inline src/plugin.py]: You can easilty add competitive playtime using the same `data-stat-id="0x0860000000000026"` like this:

```suggestion
        if game_id == "5272175":  # Overwatch
            game_time_minutes = 0
            player_data = await self.backend_client.get_ow_player_data()
            matches = re.findall('0x0860000000000026.+?(?:(?P<h>\d+):)?(?P<m>\d+)', player_data.text)
            for match in matches:
                game_time_minutes += int(match[0]) * 60 + int(match[1])
            return GameTime(game_id, game_time_minutes, None)

        return GameTime(game_id, None, None)
```

**Comment by saitho** [comment]
> This only works when your Overwatch profile is set to "public". Maybe we should add a notice to the Readme or sth.

I also thought about that. We might be able to check if the profile is private and just do nothing (if that causes an exception). Should also add a note in README.

File: src/plugin.py (hunk 1 above)
Comment (saitho) [reply src/plugin.py]: Theoretically yes. But we found out that they only show the stats for the current competitive season. There's no good way to properly fetch a total play time of all ranked seasons. :(

Comment (elkangaroo) [reply src/plugin.py]: Oh, you are right. That's too bad :(

File: src/plugin.py [hunk 2]
```diff
@@ -30,6 +30,16 @@
 
 
 class BNetPlugin(Plugin):
+    async def get_game_time(self, game_id, context):
+        qpTimeMinutes = 0
+        log.info('game_id: ' + game_id)
+        if game_id == "5272175": # Overwatch
+            player_data = await self.backend_client.get_ow_player_data()
+            qpTime = re.search('0x0860000000000026.+?(?:(?P<h>\d+):)?(?P<m>\d+)', player_data.text)
```
Comment (elkangaroo) [inline src/plugin.py]: Private profiles will still render (my profile as example: https://playoverwatch.com/en-us/career/pc/elkangaroo-2885).
In that case `re.search` will return `None`, so should check `if qpTime is not None:` in this case

**Comment by sicu** [comment]
Will this also add Overwatch to a grid/list view when sorted by 'Last played'?

File: src/plugin.py (hunk 1 above)
Comment (bartok765) [reply src/plugin.py]: is it true also for https://ow-api.com/ (only one season)?

**Comment by bartok765** [comment]
@saitho nice job. You've also proposed to use https://ow-api.com/ fandom API. Why did you decided to parse playoverwatch website?

Your profile at https://ow-api.com/v1/stats/pc/eu/saitho%232703/complete gave me json of size 100kb.
When I downloaded the https://playoverwatch.com/en-us/career/pc/saitho-2703 with `requests` and safe result as a text it has 1.4MB!

1 - 0 for ow-api.com API

File: src/plugin.py (hunk 1 above)
Comment (bartok765) [reply src/plugin.py]: @elkangaroo changes also fixes returning time for other titles: `None` should be returned for unknown value.

**Comment by bartok765** [comment]
> > This only works when your Overwatch profile is set to "public". Maybe we should add a notice to the Readme or sth.
> 
> I also thought about that. We might be able to check if the profile is private and just do nothing (if that causes an exception). Should also add a note in README.

Have you try to fetch those data in context of logged-in user? This probably will require deeper look into requests doing by playoverwatch site. I see in devtools that after login this do POST to https://identity.mparticle.com/v1/login but I don't know if it is proper track.

File: src/plugin.py (hunk 1 above)
Comment (elkangaroo) [reply src/plugin.py]: > is it true also for https://ow-api.com/ (only one season)?
yes, unfortunately it's the same.

**Comment by elkangaroo** [comment]
> > This only works when your Overwatch profile is set to "public". Maybe we should add a notice to the Readme or sth.
> 
> I also thought about that. We might be able to check if the profile is private and just do nothing (if that causes an exception). Should also add a note in README.

I found that this url returns a property "isPublic":true/false which we could use (it is a very small json response): https://playoverwatch.com/en-us/search/account-by-name/elkangaroo%232885

**Comment by saitho** [comment]
> @saitho nice job. You've also proposed to use https://ow-api.com/ fandom API. Why did you decided to parse playoverwatch website?

I opened the issue while a friend made the implementation. ^^
It might be better to use OW-API then. Will change that.

**Comment by saitho** [comment]
> Will this also correctly add Overwatch to a grid/list view when sorted by 'Last played'?

Yes :)

sort_playtime

**Comment by saitho** [comment]
> which package? Do you use virtual environment?

I figured out I had multiple Python versions installed (3.7 and 3.8). Just removed 3.8 and it works fine now. *facepalm*

I've also switched to using owapi.io, which grabs the data from the Overwatch profile and caches it there. Also the [source code is available on GitHub](https://github.com/alfg/overwatch-api) if the service was to shut down or become outdated. This still requires a public profile.

File: src/backend.py [hunk 3]
```diff
@@ -155,3 +156,11 @@ def handle_status_code(status_code):
     async def get_wow_character_achievements(self,  realm, character_name):
         url = f"https://{self._authentication_client.region}.api.blizzard.com/wow/character/{realm.lower()}/{character_name}?fields=achievements"
         return await self.do_request("GET", url)
+
+    async def get_ow_player_data(self):
+        url = 'https://owapi.io/profile/pc/' + self._authentication_client.region + '/' + self._authentication_client.user_details['battletag'].replace('#', '-')
```
Comment (bartok765) [inline src/backend.py]: Let's use f-strings for consistency. 

And didn't you forget "complete" in this url? Like
"https://ow-api.com/v1/stats/pc/eu/saitho%232703/complete" instead of "https://ow-api.com/v1/stats/pc/eu/saitho%232703"
```suggestion
        url = f"https://owapi.io/profile/pc/{self._authentication_client.region}/{self._authentication_client.user_details['battletag'].replace('#', '-')}/complete"
```

File: src/backend.py [hunk 4]
```diff
@@ -155,3 +156,11 @@ def handle_status_code(status_code):
...
+        response = await self.do_request('GET', url, None, False)
+        payload = json.loads(response.text)
```
Comment (bartok765) [inline src/backend.py]: Isn't it always json reponse? Then just:
```suggestion
        payload = await self.do_request('GET', url, json=True)
```

File: src/backend.py [hunk 5]
```diff
@@ -155,3 +156,11 @@ def handle_status_code(status_code):
...
+        if 'message' in payload: # user not found... unfortunately no 404 status code is returned :/
+            return None
```
Comment (bartok765) [inline src/backend.py]: https://ow-api.com/v1/stats/pc/eu/ssaitho%232703/complete

gives me

```json
{
    "error": "Player not found"
}
```
So probably this needs to be changed to 
```python
if 'error' in playload:
    return None
```

> user not found... unfortunately no 404 status code is returned :/

We can still raise UnknownBackendResponse or something like that in this method if you feel its more appropriate, but IMO None is OK.

File: src/plugin.py [hunk 6]
```diff
@@ -30,6 +30,28 @@
 
 
 class BNetPlugin(Plugin):
+    async def get_game_time(self, game_id, context):
```
Comment (bartok765) [inline src/plugin.py]: please move this from the top, e.g. just under get_local_games its good to have init first :)

File: src/plugin.py [hunk 7]
```diff
@@ -30,6 +30,28 @@
...
+                return GameTime(game_id, gameTimeMinutes, None)
+            qpTime = player_data['playtime']['quickplay']
+            if qpTime.count(':') == 1: # minutes and seconds
+                match = re.search('(?:(?P<m>\\d+):)(?P<s>\\d+)', qpTime)
+                gameTimeMinutes = int(qpTime.group('m'))
+            elif qpTime.count(':') == 2: # hours, minutes and seconds
+                match = re.search('(?:(?P<h>\\d+):)(?P<m>\\d+)', qpTime)
+                gameTimeMinutes = int(match.group('h')) * 60 + int(match.group('m'))
+            log.info('Overwatch quickplay playtime [minutes]: ' + str(gameTimeMinutes) + " (" + qpTime + ")")
+        return GameTime(game_id, gameTimeMinutes, None)
```
Comment (bartok765) [inline src/plugin.py]: Please do not return 0 game time for other games. Do like in @elkangaroo proposal https://github.com/bartok765/galaxy_blizzard_plugin/pull/7/files/0f1c35f92e388dceafab74bb93a6335927a8d0fb#r386485266

File: src/plugin.py [hunk 8]
```diff
@@ -30,6 +30,28 @@
...
+        log.info('game_id: ' + game_id)
+        if game_id == "5272175": # Overwatch
+            log.debug("Fetching playtime for Overwatch...")
+            player_data = await self.backend_client.get_ow_player_data()
+            if player_data == None:
+                log.error('No Overwatch profile found.')
+                return GameTime(game_id, gameTimeMinutes, None)
+            if player_data['private'] == True:
+                log.info('Unable to get data as Overwatch profile is private.')
+                return GameTime(game_id, gameTimeMinutes, None)
```
Comment (bartok765) [inline src/plugin.py]: if private - return None instead of 0 (gameTimeMinutes)

File: src/backend.py (hunk 3 above)
Comment (bartok765) [reply src/backend.py]: this suggestion need to be tested

File: src/plugin.py [hunk 9]
```diff
@@ -30,6 +30,28 @@
 
 
 class BNetPlugin(Plugin):
+    async def get_game_time(self, game_id, context):
+        gameTimeMinutes = 0
+        log.info('game_id: ' + game_id)
+        if game_id == "5272175": # Overwatch
+            log.debug("Fetching playtime for Overwatch...")
+            player_data = await self.backend_client.get_ow_player_data()
+            if player_data == None:
```
Comment (bartok765) [inline src/plugin.py]: recommended None comparison 
```suggestion
            if player_data is None:
```

File: src/plugin.py [hunk 10]
```diff
@@ -30,6 +30,28 @@
...
+                log.error('No Overwatch profile found.')
+                return GameTime(game_id, gameTimeMinutes, None)
```
Comment (bartok765) [inline src/plugin.py]: if profile not found - return None instead of 0 (gameTimeMinutes). None means "unknown"

File: src/plugin.py (hunk 7 above)
Comment (bartok765) [reply src/plugin.py]: And a minor style thing: don't use `camelCase`, but `snake_case` in variables `gameTimeMinutes`, `qpTtime`...

**Comment by bartok765** [comment]
@saitho can you also add tests (another test file)? It would be nice do it in the first place, then code changes. But if you don't know how - its OK.

File: src/plugin.py [hunk 11]
```diff
@@ -30,6 +30,28 @@
...
+            if player_data['private'] == True:
+                log.info('Unable to get data as Overwatch profile is private.')
+                return GameTime(game_id, gameTimeMinutes, None)
+            qpTime = player_data['playtime']['quickplay']
+            if qpTime.count(':') == 1: # minutes and seconds
+                match = re.search('(?:(?P<m>\\d+):)(?P<s>\\d+)', qpTime)
```
Comment (bartok765) [inline src/plugin.py]: why not using time library for this task? Regex is less obvious for me (but if you'll write good tests I'm OK with this)

File: src/backend.py (hunk 3 above)
Comment (elkangaroo) [reply src/backend.py]: > And didn't you forget "complete" in this url?

No, @saitho used https://owapi.io ([docs](https://owapi.io/docs/#api-Profile-GetProfile)) now, not https://ow-api.com - it is a different (third party) api.

File: src/backend.py (hunk 5 above)
Comment (elkangaroo) [reply src/backend.py]: The new URL now would be https://owapi.io/profile/pc/us/ssaitho-2703 which gives
`{"message":"Error: Profile not found"}`

File: src/plugin.py (hunk 11 above)
Comment (saitho) [reply src/plugin.py]: I tried: strptime is limited to a range of 0..23. The hours in playtime is not. A playtime like 100:00:00 will throw an error when processing it with strptime and %H:%M:%S

Comment (saitho) [reply src/plugin.py]: As for tests I'd like to add unit tests preferably, but I haven't found a good way of mocking the server reponse.

File: README.md [hunk 12]
```diff
@@ -1,23 +1,33 @@
...
-cd balaxy_blizard_plugin
+cd galaxy_blizzard_plugin
 pip install -r requirements/dev.txt
 inv install
 ```
 
-## Developement
+## Note on playtime
+
+Currently we only support displaying the total playtime of your quickplay matches in Overwatch. Playtime might take some time to update after you quit the game.
```
Comment (bartok765) [inline README.md]: Is it one-season only, right? We should state about this.

File: src/plugin.py (hunk 11 above)
Comment (bartok765) [reply src/plugin.py]: you can simply add json response to `tests/website_mock`.

File: README.md [hunk 13]
```diff
@@ -1,23 +1,33 @@
...
+
+**Important:** Make sure your Overwatch profile is set to public in order to show your playtime.
+To do that, start up Overwatch and navigate to the Options. From there click the "Social" tab and toggle the option "Career Profile Visibility" to "Public".
```
Comment (bartok765) [inline README.md]: Nice

File: src/plugin.py [hunk 14]
```diff
@@ -356,6 +356,29 @@ def _get_not_added_free_games(owned_games):
...
+                return GameTime(game_id, None, None)
+            qp_time = player_data['playtime']['quickplay']
+            game_time_minutes = None
+            if qp_time.count(':') == 1: # minutes and seconds
+                match = re.search('(?:(?P<m>\\d+):)(?P<s>\\d+)', qp_time)
+                game_time_minutes = int(qp_time.group('m'))
+            elif qp_time.count(':') == 2: # hours, minutes and seconds
+                match = re.search('(?:(?P<h>\\d+):)(?P<m>\\d+)', qp_time)
+                game_time_minutes = int(match.group('h')) * 60 + int(match.group('m'))
+            log.info('Overwatch quickplay playtime [minutes]: ' + str(game_time_minutes) + " (" + qp_time + ")")
```
Comment (bartok765) [inline src/plugin.py]: qp_time is already added, right?

```suggestion
            log.info(f'Overwatch quickplay playtime [minutes]: {game_time_minutes}')
```

File: README.md (hunk 12 above)
Comment (saitho) [reply README.md]: Only competitive is one season (that's why we can't add it to playtime). Quickplay is all-time and does not reset.

File: src/plugin.py (hunk 14 above)
Comment (saitho) [reply src/plugin.py]: Added tests. :) The gametime is for the entire account, not by played heroes. I covered public, private and non-existing profiles. But I have no idea how public profiles that haven't played a single game look like... I believe this to be a edge case I wouldn't cover right now. ;)

**Comment by saitho** [comment]
Profile data is bound to the battletag. During tests the battle tag is returned in mock and is always "MOCK". Haven't found a good way to adjust that in the test, which is why I'm using an environment variable to switch between different mock responses in my tests. ;)

**Comment by bartok765** [comment]
Hi, your way of changing mocks was clever, but felt it more like a hack, and I've changed those mocks to be more flexible. WebsiteClientMock was OK, but required more hacks to make it more fixture-like. Based on other plugin of FriendsOfGalaxy (steam, psn), I've created common backend_mock that is "empty" by default. 

Sync changes and use this please: 
```python
import os
import asyncio
import json
from unittest.mock import Mock

import pytest
from galaxy.api.types import GameTime

from src.backend import BackendClient
from tests.async_mock import AsyncMock


# pytest-asyncio: all test coroutines will be treated as marked.
pytestmark = pytest.mark.asyncio

OW_ID = '5272175'
NO_GAME_TIME = GameTime(OW_ID, None, None)



@pytest.fixture()
def ow_player_data_response_public():
    return json.loads("""{"username":"public","level":1307,"portrait":"https://d15f34w2p8l1cc.cloudfront.net/overwatch/182501bf4c655b69acc9ad25cf82fa0f2a8024a8d3409b09d309eab4c2239e63.png","endorsement":{"sportsmanship":{"value":0.31,"rate":31},"shotcaller":{"value":0.16,"rate":16},"teammate":{"value":0.53,"rate":53},"level":null,"frame":"https://static.playoverwatch.com/svg/icons/endorsement-frames-3c9292c49d.svg#_3","icon":"data:image/svg+xml;base64,PHN2ZyBoZWlnaHQ9IjQwIiB3aWR0aD0iNDAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiPjxjaXJjbGUgcj0iMTUuOTE1NDk0MzA5MTg5NTQiIGZpbGw9IiMyYTJiMmUiIHN0cm9rZS1kYXNoYXJyYXk9IjE2IDg0IiBzdHJva2UtZGFzaG9mZnNldD0iMjUiIHN0cm9rZS13aWR0aD0iMyIgc3Ryb2tlPSIjZjE5NTEyIiBjeD0iNTAlIiBjeT0iNTAlIj48L2NpcmNsZT48Y2lyY2xlIHI9IjE1LjkxNTQ5NDMwOTE4OTU0IiBmaWxsPSJ0cmFuc3BhcmVudCIgc3Ryb2tlLWRhc2hhcnJheT0iNTMgNDciIHN0cm9rZS1kYXNob2Zmc2V0PSIxMDkiIHN0cm9rZS13aWR0aD0iMyIgc3Ryb2tlPSIjYzgxYWY1IiBjeD0iNTAlIiBjeT0iNTAlIj48L2NpcmNsZT48Y2lyY2xlIHI9IjE1LjkxNTQ5NDMwOTE4OTU0IiBmaWxsPSJ0cmFuc3BhcmVudCIgc3Ryb2tlLWRhc2hhcnJheT0iMzEgNjkiIHN0cm9rZS1kYXNob2Zmc2V0PSI1NiIgc3Ryb2tlLXdpZHRoPSIzIiBzdHJva2U9IiM0MGNlNDQiIGN4PSI1MCUiIGN5PSI1MCUiPjwvY2lyY2xlPjx0ZXh0IHg9IjUwJSIgeT0iNTAlIiBkeT0iLjNlbSIgZm9udC1mYW1pbHk9ImNlbnR1cnkgZ290aGljLGFyaWFsLHNhbnMtc2VyaWYiIGZvbnQtd2VpZ2h0PSIzMDAiIGZvbnQtc2l6ZT0iMTYiIHN0cm9rZT0iI2Y2ZjZmNiIgc3Ryb2tlLXdpZHRoPSIxIiBmaWxsPSIjZjZmNmY2IiB0ZXh0LWFuY2hvcj0ibWlkZGxlIj5OYU48L3RleHQ+PC9zdmc+"},"private":false,"games":{"quickplay":{"won":2449},"competitive":{"won":19,"lost":21,"draw":1,"played":41,"win_rate":47.5}},"playtime":{"quickplay":"628:53:21","competitive":"07:29:56"},"competitive":{"tank":{"rank":null,"rank_img":null},"damage":{"rank":null,"rank_img":null},"support":{"rank":null,"rank_img":null}},"levelFrame":"https://d15f34w2p8l1cc.cloudfront.net/overwatch/5ab5c29e0e1e33f338ae9afc37f51917b151016aef42d10d361baac3e0965df1.png","star":"https://d15f34w2p8l1cc.cloudfront.net/overwatch/1858704e180db3578839aefdb83b89054f380fbb3d4c46b3ee12d34ed8af8712.png"}""")


@pytest.fixture()
def ow_player_data_response_private():
    return json.loads("""{"username":"private","level":452,"portrait":"https://d15f34w2p8l1cc.cloudfront.net/overwatch/e9510e804e002aa86633aeebbd158173e3432c32380a063b03f06f8f4aa401ca.png","endorsement":{"sportsmanship":{"value":0.17,"rate":17},"shotcaller":{"value":0.2,"rate":20},"teammate":{"value":0.63,"rate":63},"level":null,"frame":"https://static.playoverwatch.com/svg/icons/endorsement-frames-3c9292c49d.svg#_3","icon":"data:image/svg+xml;base64,PHN2ZyBoZWlnaHQ9IjQwIiB3aWR0aD0iNDAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiPjxjaXJjbGUgcj0iMTUuOTE1NDk0MzA5MTg5NTQiIGZpbGw9IiMyYTJiMmUiIHN0cm9rZS1kYXNoYXJyYXk9IjIwIDgwIiBzdHJva2UtZGFzaG9mZnNldD0iMjUiIHN0cm9rZS13aWR0aD0iMyIgc3Ryb2tlPSIjZjE5NTEyIiBjeD0iNTAlIiBjeT0iNTAlIj48L2NpcmNsZT48Y2lyY2xlIHI9IjE1LjkxNTQ5NDMwOTE4OTU0IiBmaWxsPSJ0cmFuc3BhcmVudCIgc3Ryb2tlLWRhc2hhcnJheT0iNjMgMzciIHN0cm9rZS1kYXNob2Zmc2V0PSIxMDUiIHN0cm9rZS13aWR0aD0iMyIgc3Ryb2tlPSIjYzgxYWY1IiBjeD0iNTAlIiBjeT0iNTAlIj48L2NpcmNsZT48Y2lyY2xlIHI9IjE1LjkxNTQ5NDMwOTE4OTU0IiBmaWxsPSJ0cmFuc3BhcmVudCIgc3Ryb2tlLWRhc2hhcnJheT0iMTcgODMiIHN0cm9rZS1kYXNob2Zmc2V0PSI0MiIgc3Ryb2tlLXdpZHRoPSIzIiBzdHJva2U9IiM0MGNlNDQiIGN4PSI1MCUiIGN5PSI1MCUiPjwvY2lyY2xlPjx0ZXh0IHg9IjUwJSIgeT0iNTAlIiBkeT0iLjNlbSIgZm9udC1mYW1pbHk9ImNlbnR1cnkgZ290aGljLGFyaWFsLHNhbnMtc2VyaWYiIGZvbnQtd2VpZ2h0PSIzMDAiIGZvbnQtc2l6ZT0iMTYiIHN0cm9rZT0iI2Y2ZjZmNiIgc3Ryb2tlLXdpZHRoPSIxIiBmaWxsPSIjZjZmNmY2IiB0ZXh0LWFuY2hvcj0ibWlkZGxlIj5OYU48L3RleHQ+PC9zdmc+"},"private":true,"games":{"quickplay":{"won":null},"competitive":{"won":null,"lost":null,"draw":0,"played":null,"win_rate":null}},"playtime":{},"competitive":{"tank":{"rank":null,"rank_img":null},"damage":{"rank":null,"rank_img":null},"support":{"rank":null,"rank_img":null}},"levelFrame":"https://d15f34w2p8l1cc.cloudfront.net/overwatch/3fdfdd16c34ab7cdc9b7be3c04197e900928b368285ce639c1d3e1c0619eea6d.png","star":"https://d15f34w2p8l1cc.cloudfront.net/overwatch/bc80149bbd78d2f940984712485bce23ddaa6f2bd0edd1c0494464ef55251eef.png"}""")


@pytest.fixture
def ow_player_data_response_not_found():
    return json.loads("""{"message":"Error: Profile not found"}""")


async def test_overwatch_private_profile(pg, backend_mock, ow_player_data_response_private):
    backend_mock.get_ow_player_data.return_value = ow_player_data_response_private

    ctx = await pg.prepare_game_times_context([OW_ID])
    assert await pg.get_game_time(OW_ID, ctx) == NO_GAME_TIME


async def test_overwatch_missing_profile(pg, backend_mock, ow_player_data_response_not_found):
    backend_mock.get_ow_player_data.return_value = ow_player_data_response_not_found

    ctx = await pg.prepare_game_times_context([OW_ID])
    assert await pg.get_game_time(OW_ID, ctx) == NO_GAME_TIME


@pytest.mark.parametrize("playtime,minutes", [
    ({'quickplay': "628:53:21", 'competitive': "07:29:56"}, (628+7)*60+53+30),
    ({'quickplay': "628:53:21", 'competitive': "0:00"}, 628*60+53),
    ({'quickplay': "0:00", 'competitive': "0:00"}, 0),  # TODO what is real response if you have never played in this season?
    # TODO test a one or two more possible edgecases
])
async def test_overwatch_public_profile(
    playtime, minutes,
    pg, backend_mock, ow_player_data_response_public
):
    ow_player_data_response_public['playtime'] = playtime
    backend_mock.get_ow_player_data.return_value = ow_player_data_response_public

    ctx = await pg.prepare_game_times_context([OW_ID])
    result = await pg.get_game_time(OW_ID, ctx)
    assert result == GameTime(OW_ID, minutes, None)

```

File: README.md (hunk 12 above)
Comment (bartok765) [reply README.md]: ah, I see

**Comment by saitho** [comment]
All done. About your question on competitive playtime: Blizzard does not reset the stats shown in the profile when a new one starts, so players will see their old stats until they start a new one.
As we don't evaluate comp time, that's not important for us.
I still added a preventive check on empty qp playtime.

One other test failed. I fixed that as well. Please have a look. :)

**Comment by saitho** [comment]
Actually I just saw an exception in logs. Will check.

File: tests/test_game_time.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
+import os
+import asyncio
+import json
+from unittest.mock import Mock
+
+import pytest
+from galaxy.api.types import GameTime
+
+from src.backend import BackendClient
```
Comment (bartok765) [inline tests/test_game_time.py]: now I see this is unused

**Comment by saitho** [comment]
> Which exception? Maybe my version is also affected.
Was an error on my part. Set a variable before declaring it. ^^

> Last test is failing on first 3 cases. You probably forget to re-run tests after this commit: [9397de2](https://github.com/bartok765/galaxy_blizzard_plugin/commit/9397de20aad3185596fd2075cdcb0979db91ff7f)

Just added your changes to this PR. Did that already fix the tests? They run fine for me. :)

**Comment by saitho** [comment]
#14 also contains changes for updating the game time. It sets the third parameter which is set here to None. If #14 gets merged first, this PR requires small adjustments. And vice versa.

**Comment by bartok765** [comment]
Yes, now it works. Let's merge!
